### DISCUSSION TITLE
Web3modal does not switch to the Metamask app on Iphone when connecting

### DISCUSSION THREAD

**Discussion by paintoshi** [WalletConnect/walletconnect-monorepo#2334] [Developer Support]
But I won't mark it as a bug because it's working as expected using the v1 example on https://lab.web3modal.com/

I just kindly request some help because I'm comparing code row against row between the source and my own implementation and can't see what I'm doing wrong?

Source: https://github.com/WalletConnect/web3modal/blob/V2/laboratory/src/pages/v1Base.tsx

My code:
![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/88572775/b0a36795-06a6-45e6-a5eb-98f076e88ecc)
The {children} in this case is pretty much the whole app given as react context providers.
The additionalChains is just fantom from 'wagmi/chains'

WalletModal.tsx
![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/88572775/e5590a7e-db0c-45ce-b5d3-6397912e22f2)

The connection flow works fine on Android, but not on iOS (iPhone 14 pro max). I just get this screen:
![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/88572775/e661f1dd-caf4-4085-b746-e0dad7443775)

Pressing open there will go to the app store. Closing the popup and use the "backup" button will switch to metamask. Sometimes metamask will let you connect but mostly nothing comes up in metamask and that doesn't work either.

What does work is the useSignMessage() from wagmi. It switches to the metamask app on both  iOS and Android.
Another thing that works is connecting a mobile wallet via the QR code option by running the app on desktop (and scanning the QR with a mobile). So 99% of the setup does work as intended.

But what does not work on any device is other wagmi functions such as writeAsync from useContractWrite(). It will NOT trigger app switching. Only in wallets such as Trust, but not with metamask.

**My installed versions:**
"react": "^18.2.0",
"@web3modal/ethereum": "^2.3.7",
"@web3modal/react": "^2.3.7",
"wagmi": "0.12.12",
"ethers": "5.7.2",

**Comment by finessevanes** [comment]
Pinging our team!

**Comment by xzilja** [comment]
That native popup "Sidan opnas..." looks like a deeplink verification one, but MetaMask doesn't have these set up, so I'm not sure why that's being shown (or is this something else, could you translate please?).

Which ios version are you on, what browser (does the issue persist in ios safari)

**Reply by paintoshi** [reply]
Yeah the "Sidan Öppnas" is just "The page will open in another app" and "Open | Cancel" buttons. Pressing open switches to the app store.

I'm on iOS 16.4.1 (a) on iPhone 14 using chrome 113.0.5672.69.
The behavior is the same on safari apart from no deeplink popup, it just goes to the app store directly.

**iOS (chrome or safari):**
Connect wallet: Works on the second try using the "backup" button
Using wagmi useSignMessage(): Swiches to metamask and back correct
Using wagmi useContractWrite(): Swiches to app store

**Android (using chrome):**
Connect wallet: Switches to metamask correct and able to connect
Using wagmi useSignMessage(): Swiches to metamask and back correctly
Using wagmi useContractWrite(): Swiches to metamask but does not send along the tx. Nothing to approve

I think it's related to this issue: https://github.com/WalletConnect/web3modal/issues/965

I have it all set up here: https://wallet-new.paintswap.finance/ if you want to test yourself. There is a !Wagmi Sign! button on the frontpage for testing purpose. I just migrated from web3.js to wagmi on the whole platform to be able to use the new web3modal, so this is the final missing piece to solve. You can go to https://wallet-new.paintswap.finance/marketplace/tools/wrapper to use the contract call method. It's on the Fantom Opera network.

**Comment by paintoshi** [comment]
Update:
After upgrading to web3modal 2.4.1 and wagmi 1.0.2 (viem 0.3.24), it started to work on iPhone. It does now switch to the metamask app when using useContractWrite(). It still opens app store when connecting to metamask, but at least it works connect using the "retry" or "backup" button on the second try.

But ran into another issue instead. It no longer auto-connect to metamask after page refresh. Using the ` autoConnect: true` for the wagmi createConfig. What's strange is it works locally running in dev mode, but not on the built app. That is when using the direct InjectedConnector from wagmi. The auto-connect still works if connecting to metamask via the web3modal.

Can be seen here (connecting directly to metamask without using the web3modal). So that's likely a wagmi issue. I still want to have the InjectedConnector as backup in case wallet connect stops working.
https://wagmi-upgrade-new.d3r28358vfuwh8.amplifyapp.com/

**Reply by xzilja** [reply]
When you say it is working with web3modal and not direct injected connector I assume you refer to mobile vs desktop injected? In that case, yeh seems like an issue on wagmi side. As for MM redirect issues, they should soon release new app version that fixes a bunch of stuff around `deeplinks` (native linking method), so as soon as that is out, we'll start using that instead, should give more stability

**Reply by paintoshi** [reply]
I'm only talking about desktop injected. But I managed to solve it by using the connector that is passed in during the wagmi config and not trying to create a new one. Mentioned here: https://github.com/wagmi-dev/wagmi/issues/2350

**Reply by xzilja** [reply]
I think I saw similar behaviour before when people initialised wagmi inside react render loops, i.e. inside page components or any other one. When you configure wagmi, are you doing all of that outside of react's scope i.e. as per docs here?https://docs.walletconnect.com/2.0/web3modal/react/installation#implementation

**Reply by paintoshi** [reply]
Pretty much yeah, I'm putting my whole app content inside that `<WagmiConfig config={wagmiConfig}></WagmiConfig>` but the config itself is outside the main render of the app.

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/88572775/cd2e5d46-710c-491f-a183-c154333048bf)

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/88572775/428f2b90-7382-4c47-a668-6311498cfc5e)

![image](https://github.com/WalletConnect/walletconnect-monorepo/assets/88572775/b0b34a68-aad2-4f54-9b26-bf4bafb2fc57)

**Reply by glitch-txs** [reply]
That's actually inside the component

Take a look at the docs, the App component and config are separated.
https://docs.walletconnect.com/2.0/web3modal/react/installation#implementation

You can move that on top of the file of your provider component to move it out of the tree
